### ISSUE TITLE
Only 2 speakers are identified by the model from the sample VCTK test data "test_1"

### ISSUE THREAD

**Issue by shobhitu** [WiraDKP/pytorch_speaker_embedding_for_diarization#3]
I had done the training of the model using the VCTK corpus and following the same structure and process. My model has trained for 29 epoch after that 'test_cost' has not minimized. 

On prediction i m only getting the following output
![Test_1_output](https://user-images.githubusercontent.com/11307998/92202716-35510600-ee9d-11ea-8d7f-e1bf471c4ea9.png)

Speakers: [0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 1 1]

What i m doing wrong?

**Comment by WiraDKP** [comment]
Hi @shobhitu,
It's okay. Just to make sure, can you share the UMAP visualization (as in Sanity Check)

However, I believe you are not in the wrong. This approach only uses window of the whole speech to diarize, so it is prone to error and still need a lot of improvement. On the other hand, we uses VCTK Dataset, which only have 109 speakers in it. That results in lack of variation to produce a good (general) speaker embedding.

For comparison, face embedding uses LFW dataset, which have around 1680 different faces, and still did not include baby and older people faces. So I think if we tried this approach on a dataset with more variation in voices, it would be better, but still not perfect.

So, why do we get different results?
Because I **sampled** the data randomly from VCTK Dataset so we may have different results.
For now, you can try to use more `n_data` in sampling the data.

Hope it helps. Thanks.

**Comment by shobhitu** [comment]
> UMAP
![Sanity_Chk](https://user-images.githubusercontent.com/11307998/92211503-30488280-eeae-11ea-80ef-5ca4e91ad0f1.png)



> Hi @shobhitu,
> It's okay. Just to make sure, can you share the UMAP visualization (as in Sanity Check)
> 
> However, I believe you are not in the wrong. This approach only uses window of the whole speech to diarize, so it is prone to error and still need a lot of improvement. On the other hand, we uses VCTK Dataset, which only have 109 speakers in it. That results in lack of variation to produce a good (general) speaker embedding.
> 
> For comparison, face embedding uses LFW dataset, which have around 1680 different faces, and still did not include baby and older people faces. So I think if we tried this approach on a dataset with more variation in voices, it would be better, but still not perfect.
> 
> So, why do we get different results?
> Because I **sampled** the data randomly from VCTK Dataset so we may have different results.
> For now, you can try to use more `n_data` in sampling the data.
> 
> Hope it helps. Thanks.

Hi thanks for the quick reply.

Sharing my UMAP generate from the Sanity Check file 
![Sanity_Chk](https://user-images.githubusercontent.com/11307998/92211624-5ff78a80-eeae-11ea-9691-1aa0a3893a82.png)


Also I had trained my model upon the same dataset having 109 speakers plus i will try to change the 'n_data' to some good no as currently we are the taking the value = 3000.

However, i have few more questions from the Training.pynb file:
1. why have to set ur callback to 15 i.e
    callback = Callback(model, config, outdir="model", early_stop_patience=15)
In my case my model has went for only 29 epoch also can we change this value to any no?

2.  Why are you tring to re-train ur training set at every 15 epoch? Is their any reason behind it?
   **while True:
    if callback.ckpt.epoch % 15 == 0:
        train_set = VCTKTripletDataset("vctk_dataset/wav48/", "vctk_dataset/txt/", n_data=3000)
        trainloader = VCTKTripletDataloader(train_set, batch_size=bs)**

why are we not using the test_set over here?

3. why have you defined the monitor == 'test_cost", since in my case test cost is not reducing much like it went to 3.4 which is quite huge. Can we put this value to 'train_cost"?
    # Early Stopping
    if callback.early_stopping(model, monitor="test_cost"):

4. What additional can be done in order to improve my model performance.

Your inputs are much appreciated..

Thanks

**Comment by WiraDKP** [comment]
The UMAP results look just fine. So there is no problem with the code I believe.

**Question 1 and 2**
Please note that I don't use the whole VCTK dataset, but I perform sampling of `n_data` and augment them as the dataset. So everytime I generate `n_data`, it would be different. It is ideal to sample and augment the data every iteration such in image classification, but it will took a long time to train. Another approach is to sample `n_data` once and train it for a while, then generate a new set of `n_data` after a few epochs. I used this approach in the code. This is just a trick to increase the data variance, while having a fast training performance.

As the data is small, the model would have early convergence. That's why I used a larger patience just to make sure it has trained enough, but remember that it would only trained on the first `n_data` generated. Then I would generate the new `n_data` every 15th epoch to push the model even further with new variation of data generated. At one point, even with a new `n_data` generated, there are no new variety, and that is when the cost would converge and the training stopped.
As you may see, number 15 would be really heuristic.

Test set should not change, so test set would only be generated once, but make sure to generate lots of it to cover a various kind of speaker.

This kind of trick won't be necessary if our data is variative enough to begin with. Unfortunately for VCTK, we only have 109 different speaker.

**Question 3**
Choosing your model should always be based on validation / test set depending your splitting scheme.
It would be prone to overfit if you decide a model based on train set.

**Question 4**
If you would like to use it for general audience, then the best solution would be to use a dataset with more speakers.
For now, try with more `n_data` and try a deeper architecture, as the one I provide is just a minimal example (6 layers CNN).
I'm using a small architecture at the time I code this to reach realtime inference.

Hope it helps.
Thanks

**Comment by shobhitu** [comment]
> you would

Thank you so much for your inputs. Will try to incorporate the following discussion

Thanks!!

**Comment by WiraDKP** [comment]
Sure. Thank you.

**Comment by shobhitu** [comment]
Hi,

I had re-trained the model by adding few extra layers and increasing the n_data = '5000', so, now i got my 'Best test_cost: 0.2106'.

But at the time of prediction i m only getting 3 speakers for 'test_1' sample with diff speaker nos, but only 2 speakers for 'test_2' and others.

Means i m still not able to meet your predicted output.


Can u please help me out in this..

Thanks!!

**Comment by WiraDKP** [comment]
If I remember, I had got 2 speakers too for `test_2.wav`. You can try to change the prediction using different clustering method. Try using different `max_frame` and `hop` too.

Also, what is the final size of your embedding?

**Comment by shobhitu** [comment]
> o, what is the final size of your embedding

Hi,

Thanks for the quick reply.

The final size of embedding that i got was 7.48 MB.

Also will try to do prediction using different parameters.

Thanks!!

**Comment by shobhitu** [comment]
Hi,

I had done the following checks in my prediction model by using different Clustering methods  and using different 'max_frame' and 'hop' too, but it doesn't change the output.

Also in the single speaker audio files model is predicting 2 speakers every time.

What is the work around for this?

Can u provide some mapping code where instead of using clustering we can calculate the mfcc score embedding for each span for the 'test_2' audio file and map those values with the model embedding and whichever gets the high score will return the speaker index for that span. 

This was my wage though, can u share some code to improve the predictability of yr model.

Thanks!!

**Comment by WiraDKP** [comment]
> The final size of embedding that i got was 7.48 MB.

No, I mean the embedding dimension that you used in your custom architecture. It should not be too small.

**Comment by WiraDKP** [comment]
> I had done the following checks in my prediction model by using different Clustering methods and using different 'max_frame' and 'hop' too, but it doesn't change the output.
> 
> Also in the single speaker audio files model is predicting 2 speakers every time.
> 
> What is the work around for this?
> 

No, the result depends on the clustering parameters. Let say if you use DBSCAN, of course it would results in more than one speaker even if there is only 1 speaker if you used a small epsilon. While using higher epsilon would results in less number of speaker clustered. I used OptimizedAHC at that time because the aim of this project is to reach realtime performance in CPU.

It is not robust to dynamically determine number of person to begin with because it is not supervised, so it really depends on the clustering. In practice, it is not common to dynamically count the number of person especially the one that is not in the dataset. What we can do is to register each person embedding, and trained a classifier using it.

On the other hand, our embedding is not that good to begin with. It is trained only on ~110 speaker in VCTK dataset. While general embedding really depends on variation in the data. With only 110 variation, I believe it would only learn high and low pitch speaker (women vs men). While similar pitch speaker have a similar embedding, hence would need a strict clustering (let say lower epsilon)

> Can u provide some mapping code where instead of using clustering we can calculate the mfcc score embedding for each span for the 'test_2' audio file and map those values with the model embedding and whichever gets the high score will return the speaker index for that span.

Mapping would mean to find similar embedding right? so more or less it would be equivalent with clustering. That's what clustering does. However, there would also a more complex mapping, such as creating a custom model to map the data using GAN. However, you would need even more data, and you may not encounter this problem when you use more variative speakers. So it's not about the number of data, it's the variety that matters.

> This was my wage though, can u share some code to improve the predictability of yr model.
> 
> Thanks!!

I have not been in audio data for a while, and I don't have time to try different ideas to. I hope I could try to improve this in the future.


I hope it helps. Sorry if it does not suit your needs.
